✨ (project) add all-in-one VM bicep deployment (#2)
DatumAcademy/iagora_deployment: README.md

# iagora_deployment

> 

Author: ## Purpose

- This PR is to facilitate the deployment of the architecture on the cloud. 
- The Bicep file includes simple modules encapsulating the complexity of the ARM file in JSON. 
- Integrating this file makes our architecture easy to understand, smooth, and reusable (with or without modification of the VM resources).

## Proposal

### Description of the Approach Taken to Achieve the Purpose

**1. **Approach and Implementation:****
   **_- **Encapsulation of Simple Modules**:_** We used simple Bicep modules to encapsulate the complexity of the ARM files in JSON. This means that each Bicep module is designed to represent a resource or a group of resources in a concise and easy-to-understand manner.
   **_- **Smooth Integration**:_** By integrating the Bicep file, we simplified the structure of our architecture. This allows for easier visualization and management of the different components of the infrastructure.
   **_- **Reusability**:_** The modular design allows the same code blocks to be reused for different deployments, with or without modifications to the VM resources. This increases efficiency and reduces the risk of errors.

**2. **Alternatives Considered:****
   **_- **Direct Use of ARM Files in JSON**:_** One alternative was to continue using ARM files in JSON directly. However, this approach is often more complex and less readable, making the management and maintenance of the infrastructure more difficult.
   **_**- **Different Infrastructure Management Tools**:_**** We also considered other infrastructure management tools like Terraform or Ansible. While these tools are powerful, we chose Bicep for its native compatibility with Azure and its ability to simplify ARM deployments.

# iagora_deployment

## Purpose

> - This PR is to facilitate the deployment of the architecture on the cloud. 

Reviewer: Why do we need the description of this PR in the README?
The README is intended to describe the project and provide documentation and guidelines for its use.
Let's postpone it for a future PR.

Author: Oh sorry! I do it now

```bash
az login --use-device-code
```

### Step 3: Install Bicep CLI

Once logged in, install Bicep CLI by running:

```bash
az bicep install
```

### Step 4: Upgrade Bicep CLI (Optional)

To ensure the latest version of Bicep is installed, upgrade it using:

```bash
az bicep upgrade
```

### Step 5: Verify Bicep Installation

Verify the Bicep installation by checking the version:

```bash
az bicep version
```

### Step 6: Deploy the Bicep Template

Since all necessary parameters are included directly within the `vm-deploy.bicep` file, there is no need to manage a separate `parameters.json` file. Deploy the resources using the following command:

```bash
az deployment group create --resource-group <your-resource-group> --template-file <path/to/template.bicep> --parameters <key1=value1> <key2=value2>
```

**Explanation of Parameters**:
- **`<your-resource-group>`**: Replace with the name of the Azure resource group where the resources will be deployed.
- **`<path/to/template.bicep>`**: Replace with the path to the Bicep file containing the deployment configuration.
- **`<key1=value1> <key2=value2>`**: Replace with custom parameters to override defaults from the Bicep file.

> ### Step 7 Customize the Deployment

Reviewer: ```suggestion
- **`<key1=value1> <key2=value2>`**: Replace with custom parameters to override defaults from the Bicep file.

### Step 7 Customize the Deployment
```
Missing newline before title?

# iagora_deployment

## How to use vm bicep deployment templates

To deploy the virtual machine and associated resources using the provided Bicep template, follow these steps:

### Step 1: Install Azure CLI

Ensure that the Azure CLI is installed. If it is not installed, use the following command:

```bash
curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash
```

### Step 2: Log in to Your Azure Account

After installing Azure CLI, log in to the Azure account:

```bash
az login
```

This command will open a browser window to log in with Azure credentials. If using a remote machine without a graphical interface, use:

```bash
az login --use-device-code
```

### Step 3: Install Bicep CLI

Once logged in, install Bicep CLI by running:

```bash
az bicep install
```

### Step 4: Upgrade Bicep CLI (Optional)

To ensure the latest version of Bicep is installed, upgrade it using:

```bash
az bicep upgrade
```

### Step 5: Verify Bicep Installation

Verify the Bicep installation by checking the version:

```bash
az bicep version
```

> ### Step 6: (Optional) Convert Existing ARM Template to Bicep
> 
> To convert an ARM template in JSON format to Bicep, use:
> 
> ```bash
> az bicep decompile -f path/to/ARM/template.json
> ```

Reviewer: ```suggestion
```
We don't provide ARM templates yet, right?

Author: right !

### Step 7: Deploy the Bicep Template

Since all necessary parameters are included directly within the `vm-deploy.bicep` file, there is no need to manage a separate `parameters.json` file. Deploy the resources using the following command:

```bash

> az deployment group create --resource-group <your-resource-group> --template-file vm-deploy.bicep

Reviewer: ```suggestion
az deployment group create --resource-group <your-resource-group> --template-file <path/to/template.bicep> --parameters <key1=value1> <key2=value2>
```

```

**Explanation of Parameters**:
- **`<your-resource-group>`**: Replace with the name of the Azure resource group where the resources will be deployed.

> - **`vm-deploy.bicep`**: This is the Bicep file containing the deployment configuration, including all required parameters. The `parameters.json` file has been removed as parameters are defined directly within this file.

Reviewer: ```suggestion
- **`<path/to/template.bicep>`**: Replace with the path to the Bicep file containing the deployment configuration.
- **`<key1=value1> <key2=value2>`**: Replace with custom parameters to override defaults from the Bicep file.
```

> ### Step 8: Customize the Deployment
> 
> Parameters can be adjusted directly in the `vm-deploy.bicep` file before running the deployment command. This approach simplifies management and customization of deployments without the need for a separate parameters file.

Reviewer: ```suggestion
```
To customize parameters, we could use the `--parameters` CLI option instead; see the suggestions above.
